### PR TITLE
 gnrc_ipv6: clean-up and simplify send handling

### PR DESCRIPTION
**Pull request by miri64**
This change is a gnrc_ipv6_nib/gnrc_netif(2)-based rework of #7210.

Packet duplication
==================

Its main optimization is that it restructures `gnrc_ipv6` handling of
sent packets so that duplication for write-protection happens at the
latest possible step:

* potential `gnrc_netif` headers added by upper layers are
  write-protected before their removal
* This unifies the duplication of the IPv6 header directly after
  that
* Extension headers in-between the IPv6 header and the payload header
  are duplicated just before the check sum is duplicated

Especially the last point allows for only handing a single packet snip
to all lower functions instead of an already searched IPv6 header
(which now is always the first until it is handed to the interface) +
payload header.

Further clean-ups
=================
* Next-hop link-layer address determination was moved to the
  `_send_unicast` function, greatly simplifying the unicast case in the
  `_send` function
* Code for loopback case was added to a new function `_send_to_self`
* Removed some code duplication

Dependencies
===========

Depends on ~~#8214~~ (merged)

### CODE REVIEW

**Comment by bergzand** [comment]
Dependency is merged (a while back).

**Comment by miri64** [comment]
The rebase isn't that trivial for doing that late and after 3 beer ^^. Will see tomorrow.

**Comment by miri64** [comment]
Rebased and fixed. There is a problem with crashes when pinging `::1`, but that also exists in master. This is fixed in #8499.

**Comment by miri64** [comment]
With #8499 I now rebased.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 1]
```diff
@@ -486,191 +536,147 @@ static void _send_multicast(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt,
...
-        /* seize payload as temporary variable */
-        ipv6 = gnrc_pktbuf_start_write(pkt); /* write protect for later removal
-                                              * in _send_unicast() */
-        if (ipv6 == NULL) {
+        /* discard broadcast and multicast flags because those could be
+         * potentially wrong (dst is later checked to assure that multicast is
+         * set if dst is a multicast address) */
+        netif_hdr_flags = netif_hdr->flags &
+                          ~(GNRC_NETIF_HDR_FLAGS_BROADCAST |
+                            GNRC_NETIF_HDR_FLAGS_MULTICAST);
```
Comment (bergzand) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Is there a place where these are set, and if so, why if they are not used anyway. Or is this just a safe guard against them being set?

**Comment by bergzand** [comment]
I can trigger a kernel panic (assertion failure) in tests/gnrc_sock_ip with this PR:

```
main(): This is RIOT! (Version: 2018.04-devel-244-g343c1-gisteren-HEAD)
Calling test_sock_ip_create__EAFNOSUPPORT()
Calling test_sock_ip_create__EINVAL_addr()
Calling test_sock_ip_create__EINVAL_netif()
Calling test_sock_ip_create__no_endpoints()
Calling test_sock_ip_create__only_local()
Calling test_sock_ip_create__only_local_reuse_ep()
Calling test_sock_ip_create__only_remote()
Calling test_sock_ip_create__full()
Calling test_sock_ip_recv__EADDRNOTAVAIL()
Calling test_sock_ip_recv__EAGAIN()
Calling test_sock_ip_recv__ENOBUFS()
Calling test_sock_ip_recv__EPROTO()
Calling test_sock_ip_recv__ETIMEDOUT()
 * Calling sock_ip_recv()
 * (timed out with timeout 1000000)
Calling test_sock_ip_recv__socketed()
Calling test_sock_ip_recv__socketed_with_remote()
Calling test_sock_ip_recv__unsocketed()
Calling test_sock_ip_recv__unsocketed_with_remote()
Calling test_sock_ip_recv__with_timeout()
Calling test_sock_ip_recv__non_blocking()
Calling test_sock_ip_send__EAFNOSUPPORT()
Calling test_sock_ip_send__EINVAL_addr()
Calling test_sock_ip_send__EINVAL_netif()
Calling test_sock_ip_send__ENOTCONN()
Calling test_sock_ip_send__socketed_no_local_no_netif()
Calling test_sock_ip_send__socketed_no_netif()
Calling test_sock_ip_send__socketed_no_local()
sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:230 => 0x5662c044
*** RIOT kernel panic:
FAILED ASSERTION.

        pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
          - | isr_stack            | -        - |   - |   8192 (   -1) | 0x5664c600 | 0x5664c600
          1 | idle                 | pending  Q |  15 |   8192 (  568) | 0x5664a420 | 0x5664c294 
          2 | main                 | running  Q |   7 |  12288 ( 3168) | 0x56647420 | 0x5664a294 
          3 | ipv6                 | bl rx    _ |   4 |   8192 ( 1344) | 0x56653540 | 0x566553b4 
            | SUM                  |            |     |  36864 ( 5080)

*** halted.
```

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 2]
```diff
@@ -340,67 +340,52 @@ static void _send_to_iface(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt)
...
-         * netif header might have been allocated by some higher layer either
-         * to set a sending interface or some flags. Interface was already
-         * copied using netif parameter, so we only need to copy the flags
-         * (minus the broadcast/multicast flags) */
-        DEBUG("ipv6: copy old interface header flags\n");
-        gnrc_netif_hdr_t *netif_new = netif_hdr->data, *netif_old = pkt->data;
-        netif_new->flags = netif_old->flags & \
-                           ~(GNRC_NETIF_HDR_FLAGS_BROADCAST | GNRC_NETIF_HDR_FLAGS_MULTICAST);
-        DEBUG("ipv6: removed old interface header\n");
-        pkt = gnrc_pktbuf_remove_snip(pkt, pkt);
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: @bergzand yes, this is actually done now immediately in [`_send()`](https://github.com/RIOT-OS/RIOT/pull/8215/files#diff-74976d45a42eaa063e3561b5fd36f615R627)

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 1 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: They are set in [`_create_netif_hdr()`](https://github.com/RIOT-OS/RIOT/pull/8215/files#diff-74976d45a42eaa063e3561b5fd36f615R358) again. Or do you mean if they may be set somewhere else? I'm not sure this is the case now, but there was the plan to use them for markers, if the checksum for UDP header compression can be elided, e.g.

**Comment by miri64** [comment]
Ping @bergzand?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 3]
```diff
@@ -448,10 +449,60 @@ static int _fill_ipv6_hdr(gnrc_netif_t *netif, gnrc_pktsnip_t *ipv6,
...
+static inline void _send_multicast_over_iface(gnrc_pktsnip_t *pkt,
+                                              gnrc_netif_t *netif,
+                                              uint8_t netif_hdr_flags)
+{
+    pkt = _create_netif_hdr(NULL, 0, pkt,
+                            netif_hdr_flags | GNRC_NETIF_HDR_FLAGS_MULTICAST);
     DEBUG("ipv6: send multicast over interface %" PRIkernel_pid "\n", netif->pid);
+    if (pkt == NULL) {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: True. Adapted it to be like the other instances `_create_netif_hdr` is called.

**Comment by bergzand** [comment]
@miri64 `tests/gnrc_sixlowpan` is broken with this PR on the samr21-xpro :(

```
2018-07-16 14:18:45,640 - INFO # main(): This is RIOT! (Version: 2018.04-devel-245-gaada6-gisteren-HEAD)
2018-07-16 14:18:45,646 - INFO # RIOT network stack example application
2018-07-16 14:18:45,650 - INFO # PKTDUMP: data received:
2018-07-16 14:18:45,656 - INFO # ~~ SNIP  0 - size:  74 byte, type: NETTYPE_SIXLOWPAN (1)
2018-07-16 14:18:45,658 - INFO # Fragmentation Header (first)
2018-07-16 14:18:45,660 - INFO # datagram size: 148
2018-07-16 14:18:45,660 - INFO # tag: 0x1
2018-07-16 14:18:45,662 - INFO # IPHC dispatch
2018-07-16 14:18:45,664 - INFO # TF: traffic class and flow label elided
2018-07-16 14:18:45,664 - INFO # NH: LOWPAN_NHC
2018-07-16 14:18:45,665 - INFO # HLIM: 255
2018-07-16 14:18:45,667 - INFO # Stateless source address compression: elided (use L2 address)
2018-07-16 14:18:45,671 - INFO # Stateless destinaton address compression: 128 bits inline
2018-07-16 14:18:45,678 - INFO # 00000000  FD  01  00  00  00  00  00  00  00  00  00  00  00  00  00  01
2018-07-16 14:18:45,684 - INFO # 00000010  F3  00  23  2F  00  00  00  00  00  00  00  00  00  00  00  00
2018-07-16 14:18:45,691 - INFO # 00000020  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
2018-07-16 14:18:45,697 - INFO # 00000030  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
2018-07-16 14:18:45,699 - INFO # 00000040  00  00  00  00
2018-07-16 14:18:45,704 - INFO # ~~ SNIP  1 - size:  26 byte, type: NETTYPE_NETIF (-1)
2018-07-16 14:18:45,707 - INFO # if_pid: 7  rssi: 0  lqi: 0
2018-07-16 14:18:45,708 - INFO # flags: 0x0
2018-07-16 14:18:45,711 - INFO # src_l2addr: 02:00:00:ff:fe:00:00:02
2018-07-16 14:18:45,714 - INFO # dst_l2addr: 02:00:00:ff:fe:00:00:01
2018-07-16 14:18:45,718 - INFO # ~~ PKT    -  2 snips, total size: 100 byte
2018-07-16 14:18:45,718 - INFO # 
2018-07-16 14:18:45,720 - INFO # Context before hardfault:
2018-07-16 14:18:45,722 - INFO #    r0: 0x00000001
2018-07-16 14:18:45,724 - INFO #    r1: 0x00000001
2018-07-16 14:18:45,725 - INFO #    r2: 0xffff0000
2018-07-16 14:18:45,727 - INFO #    r3: 0x688a881c
2018-07-16 14:18:45,728 - INFO #   r12: 0x00000000
2018-07-16 14:18:45,730 - INFO #    lr: 0x0000167f
2018-07-16 14:18:45,732 - INFO #    pc: 0x00005212
2018-07-16 14:18:45,733 - INFO #   psr: 0x21000000
2018-07-16 14:18:45,734 - INFO # 
2018-07-16 14:18:45,734 - INFO # Misc
2018-07-16 14:18:45,736 - INFO # EXC_RET: 0xfffffffd
2018-07-16 14:18:45,740 - INFO # Attempting to reconstruct state for debugging...
2018-07-16 14:18:45,741 - INFO # In GDB:
2018-07-16 14:18:45,742 - INFO #   set $pc=0x5212
2018-07-16 14:18:45,743 - INFO #   frame 0
2018-07-16 14:18:45,744 - INFO #   bt
2018-07-16 14:18:45,744 - INFO # 
2018-07-16 14:18:45,747 - INFO # ISR stack overflowed by at least 8 bytes.
```

**Comment by bergzand** [comment]
> so can you maybe see if this problem exists on master as well in the meantime).

On master the test appears to be working, it doesn't hardfault there.

**Comment by miri64** [comment]
Mh... interestingly enough I can't reproduce with master merged :-/

**Comment by miri64** [comment]
(on the branch though I can)

**Comment by bergzand** [comment]
So would another rebase on master help here then? Maybe not a bad idea to do anyway considering the age of the PR :)

**Comment by miri64** [comment]
@bergzand yes, let's try this first.

**Comment by miri64** [comment]
Rebased

**Comment by bergzand** [comment]
Looks all working to me now, I can't find any failing test at the moment. Please squash

**Comment by miri64** [comment]
Squashed

**Comment by miri64** [comment]
:tada:

**Pull request merged**